Topics/System_Design/DesignPatterns/Creational_Design_Patterns.md

```diff
@@ -0,0 +1,126 @@
+
+A singleton achieves 2 purposes: it maintains exactly 1 instance of a given class  across the entire program, and it provides a global access point to that class. It achieves this by first hiding the default constructor of the original class, and then providing a static constructor for the class. The first time the constructor is called, it generates an instance of the class and caches it within a static variable. Further calls to the constructor return the cached instance rather than creating a new one.
+
+Note that this differs from the factory method in that a singular abstract class is introduced to deal with the creation of _multiple_ related products.
+
+#### Structure
+
+The simplest implementation of a singleton design pattern consists of the desired class, adjusted to contain an instance of itself, the default constructor hidden and a getInstance method that returns the cached instance if not null, and creates a new instance(storing in the cache) if the cache is null. If the original object should not be modified, the singleton class can store it in its static cache, and should be used to get instances of the object using the getInstance method call, rather than making direct instances of the object.
+
+<p align="center">
+<img width="690" alt="UML of 2 different Singleton design patterns" src="../assets/Singleton.png">
+</p>
+
+#### Benefits
+
+- Maintain and access a central resource with a singular call to the singleton.
+
+### Builder
+
+#### Usage
+
+A highly complicated and modular object needs to be created in a step-by-step process, multiple times. Some steps may differ between iterations.
+
+A builder is a design pattern that is deployed to automate the instantiation of objects that require a step-by-step process to deal with many nested parts. It works by splitting the creation of each nested part/step into a separate method within the builder class. A director class is also introduced in order to handle the ordering/calling of these steps for a chosen builder class. By taking advantage of instantiation, it is possible to extend a given builder class or director class and override one or more methods to partially alter the creation process without having to redo the entire creation process.
+
+#### Structure
+
+The builder design pattern typically consists of an abstract builder class containing the methods(steps) common to all building processes for the given family of products, concrete implementations of that class that override/implement methods of the abstract builder, and a director class that stores a builder object and controls the calling of methods for the process.
+
+<p align="center">
+<img width="690" alt="UML of Builder design pattern" src="../assets/Builder.png">
+</p>
+
+#### Benefits
+
+- Simplifies the creation process of complicated objects.
+- Allows the adjustment of individual steps without having to rewrite the entire creation process.
+
```

Comment:
As a suggestion, adding a conclusion section summarizing each of the design patterns and comparing/contrasting them for particular scenarios that arise when programming would be a good wrap-up.